src/AzureOpenAIProxy.Management/CustomDbContextFactory.cs

```diff
@@ -0,0 +1,67 @@
+using Azure.Identity;
+using AzureOpenAIProxy.Management.Database;
+using Microsoft.EntityFrameworkCore;
+using Npgsql;
+
+namespace AzureOpenAIProxy.Management;
+
+public class CustomDbContextFactory(IConfiguration configuration) : IDbContextFactory<AoaiProxyContext>
+{
+    private readonly string? _pgConnectionString = configuration.GetConnectionString("AoaiProxyContext");
+    private readonly string? _dbHost = configuration["POSTGRES_SERVER"];
+    private readonly string _dbName = configuration["POSTGRES_DATABASE"] ?? "aoai-proxy";
+    private readonly string? _dbUser = configuration["POSTGRES_USER"];
+    private readonly string? _dbPassword = configuration["POSTGRES_PASSWORD"];
+    private DbContextOptionsBuilder<AoaiProxyContext>? _pgOptionsBuilder = null;
+    private NpgsqlDataSourceBuilder? _pgDataSourceBuilder = null;
+    private NpgsqlDataSource? _pgDataSource = null;
+    private DateTime _connectionTime = DateTime.MinValue;
+    private const int _maxConnectionTime = 60 * 4; // 4 hrs
+
+    async private Task<string> GetConnectionString()
+    {
+        if (!string.IsNullOrEmpty(_pgConnectionString))
+        {
+            Console.WriteLine("Using Postgres Connection String");
+            return _pgConnectionString;
+        }
+
+        if (!string.IsNullOrEmpty(_dbPassword))
+        {
+            Console.WriteLine("Using Postgres Native Authorization");
+            return $"Host={_dbHost};Database={_dbName};Username={_dbUser};Password={_dbPassword}";
+        }
+        else
+        {
+            Console.WriteLine("Using Postgres Entra Authorization");
+            var sqlServerTokenProvider = new DefaultAzureCredential();
+            string accessToken = (await sqlServerTokenProvider.GetTokenAsync(
+                 new Azure.Core.TokenRequestContext(scopes: new string[] { "https://ossrdbms-aad.database.windows.net/.default" }) { })).Token;
```

Comment:
```suggestion
            // only using `var` because I'm not sure the real type here
            var tokenResponse = await sqlServerTokenProvider.GetTokenAsync(
                 new Azure.Core.TokenRequestContext(scopes: new string[] { "https://ossrdbms-aad.database.windows.net/.default" }) { });
            string accessToken = tokenResponse.Token;
```

Let's split the calls into two steps rather than wrapping in `()`, just for readability.